### ISSUE TITLE
twisted.protocols.tls uses quadratic complexity buffer management when writing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6477]
|<img alt="hud's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| hud reported|
|-|-|
|Trac ID|trac#6477|
|Type|defect|
|Created|2013-04-26 01:15:22Z|
|Branch|https://github.com/twisted/twisted/tree/ssl-largewrite-throughput-6477|

Hi,

 I'm new to this code. I'm sending 40 MB via TLS, and _write(self, bytes) in twisted/protocols/tls.py uses

 leftToSend = leftToSend[sent:]

 which is extremely slow since 16kb is written each time. It takes half a minute to send this 40 MB and uses 100% of one of the cores.

 I modified the code in order to send only bytes[alreadySent:alreadySent+16384], where alreadySent += sent. This solved the issue. Sorry for not attaching diff, I don't have time, I just wanted to let you guys know.

 Peter

<details><summary>Searchable metadata</summary>

```
trac-id__6477 6477
type__defect defect
reporter__hud hud
priority__normal normal
milestone__ 
branch__branches_ssl_largewrite_throughput_6477 branches/ssl-largewrite-throughput-6477
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1366938922000000 1366938922000000
changetime__1367322144000000 1367322144000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [38274]) Branching to 'ssl-largewrite-throughput-6477'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [38275]) Walk a buffer through the input instead of repeatedly slicing the unwritten part off the end

refs [#6477](https://github.com/twisted/twisted/issues/6477)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I think I dealt with this in the branch.

Here are the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ssl-largewrite-throughput-6477).  Unfortunately the slave for the Python 3 builder is MIA.  I'm not comfortable merging until that comes back and shows us this works there as well.  Perhaps it will by the time a reviewer gets here, though.

I also added two new benchmarks to lp:twisted-benchmarks, currently [proposed for merging on Launchpad](https://code.launchpad.net/~exarkun/twisted-benchmarks/more-ssl-benchmarks/+merge/161113).  It'd be nice to merge and deploy those changes first, so we can see things actually get faster.

For a taste, on my 2.2GHz Celeron laptop, I observed no change in ssl_throughput and a 30% speedup in ssl_throughput_bigwrite and ssl_throughput_smallwrite.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Actually, scratch the bit about ssl_throughput_smallwrite - I don't think it changes much either.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

It looks good to me. I'd just mention that no test fail if I replace `alreadySend` by 0 in the `except WantReadError` clause. It also makes me notice that we do a useless copy here when alreadySend is 0 (probably very minor).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [38313]) Merge ssl-largewrite-throughput-6477

Author: exarkun
Reviewer: therve
Fixes: [#6477](https://github.com/twisted/twisted/issues/6477)

In `twisted.protocols.tls`, replace some O(N**2) complexity buffer handling
code with some code of somewhat lower complexity.  This primarily helps
improve performance when a single call to `write` is passed a very large
string (the further above 2 ** 16, the bigger impact this will likely
have).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> I'd just mention that no test fail if I replace alreadySend by 0 in the except WantReadError clause. 

Filed [#6500](https://github.com/twisted/twisted/issues/6500) for this.

> It also makes me notice that we do a useless copy here when alreadySend is 0 (probably very minor).

CPython does this micro-optimization for us - `s[0:] is s`.  I assume PyPy is smart enough not to waste time doing a copy for this operation, too.
